design-documents/testing/functional/bdd-like-mftf-output.md

```diff
@@ -0,0 +1,141 @@
+- Allure Report
+- CLI Report
+- Documentation
+
+## Behavioral output
+
+The real value of such output is possibility to confront the Business Requirements with Test Scenario without any extra interpretation. The content should be obvious to non-technical person.
+
+Example of such is Sylius Framework: https://github.com/Sylius/Sylius/blob/48729b86334ba78d8f2011b9743389024c334d56/features/checkout/paying_for_order/paying_with_paypal_during_checkout.feature#L7-L24
+
+### User-defined Label Attribute
+
+Adding extra attribute to XML nodes that allows to define user-friendly label for element / action / action group.
+
+* Element
+ ```xml
+ <element name="password" type="input" selector="#login" label="Password Input"/>
+ ```
+* Performed Action
+  ```xml
+  <click selector=".niceButton" stepKey="clickButton" label="Click Nice Button"/>
+  ```
+* Action Group
+  ```xml
+  <actionGroup name="resetProductGridToDefaultView" label="Reset Product Grid">
+  ```
+* Action Group Execution
+  ```xml
+  <actionGroup ref="someActionGroup" label="Running Action Group">
+  ```
+
+#### Benefits
+
+* MFTF user is able to provide domain-consistent labels.
+
+### Generated Label
+
+* Element
+  ```xml
+  <element name="password" type="input" selector="#login"/>
+  ```
+  generates:
+  ```gherkin
+  Password input
+  ```
+* Action Group
+  ```xml
+  <actionGroup name="StorefrontFillCustomerAccountCreationFormActionGroup">
+      <arguments>
+          <argument name="customer" type="entity" />
+      </arguments>
+
+      <fillField  stepKey="fillFirstName" userInput="{{customer.firstname}}" selector="{{StorefrontCustomerCreateFormSection.firstnameField}}" />
+      <fillField  stepKey="fillLastName"  userInput="{{customer.lastname}}" selector="{{StorefrontCustomerCreateFormSection.lastnameField}}" />
+      <fillField  stepKey="fillEmail" userInput="{{customer.email}}" selector="{{StorefrontCustomerCreateFormSection.emailField}}"/>
+      <fillField  stepKey="fillPassword" userInput="{{customer.password}}" selector="{{StorefrontCustomerCreateFormSection.passwordField}}"/>
+      <fillField  stepKey="fillConfirmPassword" userInput="{{customer.password}}" selector="{{StorefrontCustomerCreateFormSection.confirmPasswordField}}"/>
+  </actionGroup>
+  ```
+  generates:
+  ```gherkin
+  Action group summary:
+   - Storefront Fill Customer Account Creation Form
+
+  Requires arguments:
+   - 'customer' entity
+
+  Steps:
+  1. Fill First Name
+  2. Fill Last Name
+  3. Fill Email
+  4. Fill Password
+  5. Fill Confirm Password
+  ```
+
+#### Benefits
+
+* Automatic label generation enforces MFTF user to provide descriptive `name`s / `stepKey`s in Test Scenarios
+
+# Implementation
+
+1. **MFTF Naming Rules**
+
+  There's no standard regarding naming in MFTF Tests. One should be formed and published in Magento DevDocs, as well as in the MFTF repository. Tests migrated / created after forming such rules should follow the standards.
```

Comment:
As far as I understand, this is currently in progress (@tomreece ?)

Reply:
Yup! We are currently incrementally adding content to the heading "Functional Testing" on this page:
https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html#113-functional-testing

These rules were carefully groomed by the entire Pangolins team and should be obeyed by all.